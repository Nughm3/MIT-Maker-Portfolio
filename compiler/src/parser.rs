use cstree::{
    build::{Checkpoint, GreenNodeBuilder, NodeCache},
    interning::Interner,
    text::TextRange,
};

use crate::syntax::*;

mod grammar;

#[cfg(test)]
mod tests;

const ASSIGNMENT_OPS: &[SyntaxKind] = &[
    SyntaxKind::Equals,
    SyntaxKind::PlusEquals,
    SyntaxKind::MinusEquals,
    SyntaxKind::StarEquals,
    SyntaxKind::SlashEquals,
    SyntaxKind::PercentEquals,
];

const TRIVIA_TOKENS: &[SyntaxKind] = &[
    SyntaxKind::Comment,
    SyntaxKind::Newlines,
    SyntaxKind::Whitespace,
];

#[derive(Debug)]
pub struct ParseError {
    pub expected: Vec<SyntaxKind>,
    pub found: SyntaxKind,
    pub range: TextRange,
}

pub fn parse(
    input: &str,
    node_cache: &mut NodeCache<'_, impl Interner>,
) -> (SyntaxNode, Vec<ParseError>) {
    let mut parser = Parser::new(input, node_cache);
    grammar::file(&mut parser);
    parser.finish()
}

struct Parser<'input, 'cache, 'interner, I> {
    input: &'input str,
    tokens: Vec<SyntaxKind>,
    ranges: Vec<TextRange>,
    pos: usize,
    errors: Vec<ParseError>,
    builder: GreenNodeBuilder<'cache, 'interner, SyntaxKind, I>,
}

impl<'input, 'cache, 'interner, I: Interner> Parser<'input, 'cache, 'interner, I> {
    fn new(input: &'input str, node_cache: &'cache mut NodeCache<'interner, I>) -> Self {
        let (tokens, ranges) = SyntaxKind::lexer(input).unzip();

        Parser {
            input,
            tokens,
            ranges,
            pos: 0,
            errors: Vec::new(),
            builder: GreenNodeBuilder::with_cache(node_cache),
        }
    }

    fn finish(self) -> (SyntaxNode, Vec<ParseError>) {
        (SyntaxNode::new_root(self.builder.finish().0), self.errors)
    }

    fn start_node(&mut self, kind: SyntaxKind) {
        self.builder.start_node(kind);
    }

    fn start_node_at(&mut self, checkpoint: Checkpoint, kind: SyntaxKind) {
        self.builder.start_node_at(checkpoint, kind);
    }

    fn finish_node(&mut self) {
        self.builder.finish_node();
    }

    fn checkpoint(&self) -> Checkpoint {
        self.builder.checkpoint()
    }

    fn skip_trivia(&mut self) {
        while let Some(kind) = self.tokens.get(self.pos).copied() {
            if !TRIVIA_TOKENS.contains(&kind) {
                break;
            }

            let text = &self.input[self.ranges[self.pos]];
            self.builder.token(kind, text);
            self.pos += 1;
        }
    }

    fn next(&mut self) -> SyntaxKind {
        self.skip_trivia();

        if let Some(kind) = self.tokens.get(self.pos).copied() {
            let span = self.ranges[self.pos];
            let text = &self.input[span];
            self.builder.token(kind, text);
            self.pos += 1;
            kind
        } else {
            SyntaxKind::Eof
        }
    }

    fn peek(&self) -> SyntaxKind {
        self.tokens
            .get(self.pos..)
            .and_then(|tokens| tokens.iter().find(|token| !TRIVIA_TOKENS.contains(*token)))
            .copied()
            .unwrap_or(SyntaxKind::Eof)
    }

    fn at(&self, kind: SyntaxKind) -> bool {
        self.peek() == kind
    }

    fn at_any(&self, kinds: &[SyntaxKind]) -> bool {
        kinds.contains(&self.peek())
    }

    fn eof(&self) -> bool {
        self.at(SyntaxKind::Eof)
    }

    fn consume(&mut self, kind: SyntaxKind) -> bool {
        let consumed = self.at(kind);
        if consumed {
            self.next();
        }
        consumed
    }

    fn expect(&mut self, expected: SyntaxKind) -> bool {
        let consumed = self.consume(expected);
        if !consumed {
            self.next_error(&[expected]);
        }
        consumed
    }

    fn next_error(&mut self, expected: &[SyntaxKind]) {
        self.start_node(SyntaxKind::ErrorTree);
        let found = self.next();
        self.errors.push(ParseError {
            expected: expected.to_vec(),
            found,
            range: self.ranges.get(self.pos - 1).copied().unwrap_or_default(),
        });
        self.finish_node();
    }

    fn recover_to(&mut self, recovery: &[SyntaxKind]) {
        self.start_node(SyntaxKind::ErrorTree);
        while !self.at_any(recovery) && !self.eof() {
            self.next();
        }
        self.finish_node();
    }
}
