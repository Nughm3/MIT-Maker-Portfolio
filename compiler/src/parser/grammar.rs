use SyntaxKind::*;

use super::*;

// Inline grammar tests
// - use "//@" to specify the rule to test
// - use "//#" to create a new test
// - use "//-" to continue the last test onto the next line

const ITEM_FIRST: &[SyntaxKind] = &[ImportKw, TypeKw, FnKw, ConstKw];

const STMT_FIRST: &[SyntaxKind] = &[
    Semicolon, LetKw, LoopKw, WhileKw, BreakKw, ContinueKw, ReturnKw,
];

const EXPR_FIRST: &[SyntaxKind] = &[
    LeftBrace, // Block
    IfKw,      // ExprIf
    LeftParen, // ExprParen
    Minus,     // ExprPrefix
    NotKw,     // ExprPrefix
    FnKw,      // ExprClosure
    True,      // 'true'
    False,     // 'false'
    Int,       // 'int'
    Float,     // 'float'
    Char,      // 'char'
    String,    // 'string'
    Ident,     // Path
];

//@ file
//# fn f() {}
//- const x = 10 : Int;
pub fn file(p: &mut Parser<'_, '_, '_, impl Interner>) {
    p.start_node(File);

    while !p.eof() {
        if p.at_any(ITEM_FIRST) {
            item(p);
        } else {
            p.next_error(ITEM_FIRST);
            p.recover_to(ITEM_FIRST);
        }
    }

    p.finish_node();
}

pub fn item(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert!(p.at_any(ITEM_FIRST));
    match p.peek() {
        ImportKw => import(p),
        TypeKw => type_def(p),
        FnKw => function(p),
        ConstKw => constant(p),
        _ => p.next_error(ITEM_FIRST),
    }
}

fn nested_item(p: &mut Parser<'_, '_, '_, impl Interner>) {
    if p.at_any(&[FnKw, ConstKw]) {
        match p.peek() {
            FnKw => function(p),
            ConstKw => constant(p),
            _ => unreachable!(),
        }
    } else {
        p.start_node(ErrorTree);
        p.next_error(&[FnKw, ConstKw]);
        p.finish_node();
    }
}

//@ import
//# import foo;
//# import foo.bar;
pub fn import(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), ImportKw);
    p.start_node(Import);
    p.next();
    path(p);
    p.expect(Semicolon);
    p.finish_node();
}

//@ type_def
//# type Foo(bar: Bar, baz: Baz)
pub fn type_def(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), TypeKw);
    p.start_node(TypeDef);
    p.next();
    p.expect(Ident);

    p.expect(LeftParen);
    while !p.at(RightParen) && !p.eof() {
        p.start_node(AdtField);
        p.expect(Ident);
        p.expect(Colon);
        type_expr(p);
        if p.at(RightParen) {
            p.consume(Comma);
        } else {
            p.expect(Comma);
        }
        p.finish_node();
    }
    p.expect(RightParen);
    p.finish_node();
}

//@ type_expr
//# fn()
//# fn(Int, Int) -> Int
//# Int
pub fn type_expr(p: &mut Parser<'_, '_, '_, impl Interner>) {
    match p.peek() {
        FnKw => {
            p.start_node(TypeFunction);
            p.next();
            p.expect(LeftParen);
            while !p.at(RightParen) && !p.eof() {
                type_list(p);
            }
            p.expect(RightParen);
            if p.consume(Arrow) {
                type_expr(p);
            }
            p.finish_node();
        }
        Ident => path(p),
        _ => p.next_error(&[FnKw, Ident]),
    }
}

fn type_list(p: &mut Parser<'_, '_, '_, impl Interner>) {
    p.start_node(TypeList);
    type_expr(p);
    if p.at(RightParen) {
        p.consume(Comma);
    } else {
        p.expect(Comma);
    }
    p.finish_node();
}

//@ function
//# fn f() -> Int { return 10; }
pub fn function(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), FnKw);
    p.start_node(Function);
    signature(p);
    if p.at(LeftBrace) {
        block(p);
    }
    p.finish_node();
}

//@ signature
//# fn f(x: Int) -> Int
pub fn signature(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), FnKw);
    p.start_node(Signature);
    p.next();
    p.expect(Ident);
    p.expect(LeftParen);
    while !p.at(RightParen) && !p.eof() {
        param_list(p);
    }
    p.expect(RightParen);
    if p.consume(Arrow) {
        type_expr(p);
    }
    p.finish_node();
}

fn param_list(p: &mut Parser<'_, '_, '_, impl Interner>) {
    p.start_node(ParamList);
    p.expect(Ident);
    if p.consume(Colon) {
        type_expr(p);
    }
    if p.at(RightParen) {
        p.consume(Comma);
    } else {
        p.expect(Comma);
    }
    p.finish_node();
}

//@ constant
//# const x = 5;
//# const x = 5 : Int;
pub fn constant(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), ConstKw);
    p.start_node(Constant);
    p.next();
    p.expect(Ident);
    p.expect(Equals);
    if p.at_any(EXPR_FIRST) {
        expr(p);
    }
    if p.at(Colon) {
        p.start_node(TypeAscription);
        p.next();
        type_expr(p);
        p.finish_node();
    }
    p.expect(Semicolon);
    p.finish_node();
}

//@ block
//# {}
//# { a; b; c; }
pub fn block(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), LeftBrace);
    p.start_node(Block);
    p.expect(LeftBrace);
    while !p.at(RightBrace) && !p.eof() {
        let expected = &[STMT_FIRST, EXPR_FIRST, ITEM_FIRST].concat();
        if p.at_any(expected) {
            stmt(p);
        } else {
            p.next_error(expected);
        }
    }
    p.expect(RightBrace);
    p.finish_node();
}

//@ stmt
//# ;
//# a = 5;
//# break;
//# continue;
//# return;
//# return x;
//# fn f() {}
//# const x = 5;
pub fn stmt(p: &mut Parser<'_, '_, '_, impl Interner>) {
    match p.peek() {
        LetKw => stmt_let(p),
        LoopKw => stmt_loop(p),
        WhileKw => stmt_while(p),
        BreakKw => {
            p.start_node(StmtBreak);
            p.next();
            p.expect(Semicolon);
            p.finish_node();
        }
        ContinueKw => {
            p.start_node(StmtContinue);
            p.next();
            p.expect(Semicolon);
            p.finish_node();
        }
        ReturnKw => {
            p.start_node(StmtReturn);
            p.next();
            if !p.at(Semicolon) {
                expr(p);
            }
            p.expect(Semicolon);
            p.finish_node();
        }
        Semicolon => {
            p.next();
        }
        FnKw | ConstKw => nested_item(p),
        e if EXPR_FIRST.contains(&e) => {
            p.start_node(StmtExpr);
            if let ExprKind::Inline = expr(p) {
                p.expect(Semicolon);
            }
            p.finish_node();
        }
        // NOTE: FnKw is in EXPR_START due to ExprClosure
        _ => p.next_error(&[STMT_FIRST, EXPR_FIRST, &[ConstKw]].concat()),
    }
}

//@ stmt_let
//# let x = 5;
//# let x = 5 : Int;
pub fn stmt_let(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), LetKw);
    p.start_node(StmtLet);
    p.next();
    p.expect(Ident);
    p.expect(Equals);
    if p.at_any(EXPR_FIRST) {
        expr(p);
    }
    if p.at(Colon) {
        p.start_node(TypeAscription);
        p.next();
        type_expr(p);
        p.finish_node();
    }
    p.expect(Semicolon);
    p.finish_node();
}

//@ stmt_loop
//# loop { continue; }
pub fn stmt_loop(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), LoopKw);
    p.start_node(StmtLoop);
    p.next();
    if p.at(LeftBrace) {
        block(p);
    }
    p.finish_node();
}

//@ stmt_while
//# while x < 10 { x += 1; }
pub fn stmt_while(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), WhileKw);
    p.start_node(StmtWhile);
    p.next();
    if p.at_any(EXPR_FIRST) {
        expr(p);
    }
    if p.at(LeftBrace) {
        block(p);
    }
    p.finish_node();
}

#[derive(Debug)]
pub enum ExprKind {
    Inline,
    Block,
}

#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
struct BindingPower(u8);

impl BindingPower {
    const MIN: Self = BindingPower(u8::MIN);
    const MAX: Self = BindingPower(u8::MAX);

    fn of(op: SyntaxKind) -> Option<(Self, Self)> {
        let bp = [
            ASSIGNMENT_OPS,
            &[OrKw],
            &[AndKw],
            &[Eq, Ne, Lt, Le, Gt, Ge],
            &[Plus, Minus],
            &[Star, Slash, Percent],
        ]
        .iter()
        .position(|level| level.contains(&op))
        .map(|bp| bp as u8 * 2)?;

        Some(if ASSIGNMENT_OPS.contains(&op) {
            (BindingPower(bp + 2), BindingPower(bp + 1))
        } else {
            (BindingPower(bp + 1), BindingPower(bp + 2))
        })
    }
}

//@ expr
//# (a)
//# -1
//# not false
//# x + y
//# x + y * z
//# (x + y) * z
//# f()
//# f(x, y, z)
//# f()()
//# Foo(bar=baz)
//# fn() x
//# fn(x) x
//# fn(x: Int) -> Int 2 * x
//# fn(x, y, z) { return x + y + z; }
//# true
//# false
//# 123
//# 12_3
//# 0xff
//# 0xFF
//# 0b1111_0000
//# 1.23
//# 1.
//# 1e5
//# 1.0E-2_3
//# 'x'
//# '\''
//# '\0'
//# ""
//# "hello world"
//# "this is a quote: \""
//# i_dentifier
//# path.literal
pub fn expr(p: &mut Parser<'_, '_, '_, impl Interner>) -> ExprKind {
    expr_rec(p, BindingPower::MIN)
}

fn expr_rec(p: &mut Parser<'_, '_, '_, impl Interner>, bp: BindingPower) -> ExprKind {
    let checkpoint = p.checkpoint();

    let mut kind = match p.peek() {
        LeftBrace => {
            block(p);
            ExprKind::Block
        }
        IfKw => {
            expr_if(p);
            ExprKind::Block
        }
        LeftParen => {
            p.start_node(ExprParen);
            p.next();
            expr(p);
            p.expect(RightParen);
            p.finish_node();
            ExprKind::Inline
        }
        Minus | NotKw => {
            p.start_node(ExprPrefix);
            p.next();
            expr_rec(p, BindingPower::MAX);
            p.finish_node();
            ExprKind::Inline
        }
        FnKw => {
            p.start_node(ExprClosure);
            p.next();
            p.expect(LeftParen);
            while !p.at(RightParen) && !p.eof() {
                param_list(p);
            }
            p.expect(RightParen);
            if p.consume(Arrow) {
                type_expr(p);
            }
            let kind = expr(p);
            p.finish_node();
            kind
        }
        True | False | Int | Float | Char | String => {
            p.next();
            ExprKind::Inline
        }
        Ident => {
            path(p);
            ExprKind::Inline
        }
        _ => {
            p.next_error(EXPR_FIRST);
            ExprKind::Inline
        }
    };

    loop {
        match p.peek() {
            LeftParen => {
                p.start_node_at(checkpoint, ExprCall);
                p.next();
                while !p.at(RightParen) && !p.eof() {
                    p.start_node(ExprList);
                    expr(p);
                    if p.at(RightParen) {
                        p.consume(Comma);
                    } else {
                        p.expect(Comma);
                    }
                    p.finish_node();
                }
                p.expect(RightParen);
                p.finish_node();
            }
            op => {
                if let Some((l, r)) = BindingPower::of(op) {
                    if l < bp {
                        break;
                    }

                    p.start_node_at(
                        checkpoint,
                        if ASSIGNMENT_OPS.contains(&op) {
                            ExprAssign
                        } else {
                            ExprInfix
                        },
                    );

                    p.next();
                    expr_rec(p, r);
                    p.finish_node();
                } else {
                    break;
                }
            }
        }

        kind = ExprKind::Inline;
    }

    kind
}

//@ expr_if
//# if x { y; }
//# if x { y; } else { z; }
//# if x {} else if y {}
//# if x {} else if y {} else if z {} else {}
pub fn expr_if(p: &mut Parser<'_, '_, '_, impl Interner>) {
    debug_assert_eq!(p.peek(), IfKw);
    p.start_node(ExprIf);
    p.next();
    expr(p);
    if p.at(LeftBrace) {
        block(p);
    }
    if p.at(ElseKw) {
        p.next();
        match p.peek() {
            IfKw => expr_if(p),
            LeftBrace => block(p),
            _ => p.next_error(&[IfKw, LeftBrace]),
        }
    }
    p.finish_node();
}

//@ path
//# foo
//# bar.baz
//# package
//# package.foo.bar
pub fn path(p: &mut Parser<'_, '_, '_, impl Interner>) {
    p.start_node(Path);
    if !p.consume(PackageKw) {
        p.start_node(PathComponent);
        p.expect(Ident);
        p.finish_node();
    }
    while p.at(Dot) && !p.eof() {
        p.start_node(PathComponent);
        p.next();
        p.expect(Ident);
        p.finish_node();
    }
    p.finish_node();
}
