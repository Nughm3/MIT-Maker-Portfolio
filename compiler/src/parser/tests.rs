use std::fmt::Write;

use ahash::{HashMap, HashMapExt};
use cstree::build::NodeCache;

use super::{grammar::*, ParseError, Parser, SyntaxNode};
use crate::parser::SyntaxKind;

// TODO: generate tests with build script or macro, test AST layer

const GRAMMAR: &str = include_str!("grammar.rs");

#[test]
fn parser() {
    let mut test_suite: HashMap<_, Vec<_>> = HashMap::new();
    let mut context = String::new();

    for line in GRAMMAR.lines().filter(|l| l.starts_with("//")) {
        let content = line[4..].to_string();
        match &line[2..3] {
            "@" => context = content,
            "#" => test_suite.entry(context.clone()).or_default().push(content),
            "-" => {
                let test = test_suite
                    .entry(context.clone())
                    .or_default()
                    .last_mut()
                    .expect("cannot continue in an empty test group");
                test.push('\n');
                test.push_str(&content);
            }
            _ => continue,
        }
    }

    let mut cache = NodeCache::new();
    for (group, tests) in test_suite.iter() {
        for (idx, test) in tests.iter().enumerate() {
            eprintln!("{group}#{idx}");
            let (cst, parse_errors, unused_tokens) = parse(group, test, &mut cache);

            let format_cst = |cst: SyntaxNode| {
                cst.debug(cache.interner(), true)
                    .lines()
                    .skip(1)
                    .fold(String::new(), |mut s, l| {
                        let _ = writeln!(s, "{}", &l[2..]);
                        s
                    })
            };

            let mut output = String::new();

            if !parse_errors.is_empty() || !unused_tokens.is_empty() {
                output.push_str(&parse_errors.into_iter().fold(String::new(), |mut s, e| {
                    let _ = writeln!(
                        s,
                        "ERROR: expected {:?}, found {:?} ({:?})",
                        e.expected, e.found, e.range
                    );
                    s
                }));

                output.push_str(&unused_tokens.into_iter().fold(String::new(), |mut s, t| {
                    let _ = writeln!(s, "UNUSED: {t:?}");
                    s
                }));

                output.push_str("---\n");
            }

            output.push_str(&format_cst(cst));

            insta::with_settings!({
                omit_expression => true,
                description => test,
            }, {
                insta::assert_snapshot!(format!("{group}#{idx}"), output);
            });
        }
    }
}

fn parse(
    rule: &str,
    input: &str,
    cache: &mut NodeCache<'static>,
) -> (SyntaxNode, Vec<ParseError>, Vec<SyntaxKind>) {
    let mut parser = Parser::new(input, cache);
    parser.start_node(SyntaxKind::ParserTest);

    match rule {
        "file" => file(&mut parser),
        "item" => item(&mut parser),
        "import" => import(&mut parser),
        "type_def" => type_def(&mut parser),
        "type_expr" => type_expr(&mut parser),
        "function" => function(&mut parser),
        "signature" => signature(&mut parser),
        "constant" => constant(&mut parser),
        "block" => block(&mut parser),
        "stmt" => stmt(&mut parser),
        "stmt_let" => stmt_let(&mut parser),
        "stmt_loop" => stmt_loop(&mut parser),
        "stmt_while" => stmt_while(&mut parser),
        "expr" => {
            expr(&mut parser);
        }
        "expr_if" => expr_if(&mut parser),
        "path" => path(&mut parser),
        err => unreachable!("rule {err} not added to parse()"),
    }

    parser.finish_node();
    let unused_tokens = parser.tokens[parser.pos..].to_vec();
    let (cst, parse_errors) = parser.finish();
    (cst, parse_errors, unused_tokens)
}
