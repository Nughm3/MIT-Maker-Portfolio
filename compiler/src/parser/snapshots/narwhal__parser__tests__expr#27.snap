---
source: src/parser/tests.rs
description: "'\\0'"
---
Char@0..4 "'\\0'"
